cpp/src/arrow/compute/kernels/filter.h

```diff
@@ -32,120 +33,45 @@ namespace compute {
-              const ChunkedArray& filter, std::shared_ptr<ChunkedArray>* out);
-
-/// \brief Filter a record batch with a boolean selection filter
-///
-/// The output record batch's columns will be populated with values from corresponding
-/// columns of the input at positions where the selection filter is not 0. Nulls in the
-/// filter will result in nulls in the output.
-///
-/// \param[in] ctx the FunctionContext
-/// \param[in] batch record batch to filter
-/// \param[in] filter indicates which values should be filtered out
-/// \param[out] out resulting record batch
-/// NOTE: Experimental API
-ARROW_EXPORT
-Status Filter(FunctionContext* ctx, const RecordBatch& batch, const Array& filter,
-              std::shared_ptr<RecordBatch>* out);
-
-/// \brief Filter a table with a boolean selection filter
-///
-/// The output table's columns will be populated with values from corresponding
-/// columns of the input at positions where the selection filter is not 0. Nulls in the
-/// filter will result in nulls in each column of the output.
-///
-/// \param[in] ctx the FunctionContext
-/// \param[in] table table to filter
-/// \param[in] filter indicates which values should be filtered out
-/// \param[out] out resulting table
-/// NOTE: Experimental API
-ARROW_EXPORT
-Status Filter(FunctionContext* ctx, const Table& table, const Array& filter,
-              std::shared_ptr<Table>* out);
-
-/// \brief Filter a table with a boolean selection filter
-///
-/// The output record batch's columns will be populated with values from corresponding
-/// columns of the input at positions where the selection filter is not 0. Nulls in the
-/// filter will result in nulls in the output.
-///
-/// \param[in] ctx the FunctionContext
-/// \param[in] table record batch to filter
-/// \param[in] filter indicates which values should be filtered out
-/// \param[out] out resulting record batch
-/// NOTE: Experimental API
-ARROW_EXPORT
-Status Filter(FunctionContext* ctx, const Table& table, const ChunkedArray& filter,
-              std::shared_ptr<Table>* out);
-
-/// \brief Filter an array with a boolean selection filter
-///
-/// \param[in] ctx the FunctionContext
-/// \param[in] values datum to filter
+/// \param[in] values array to filter
 /// \param[in] filter indicates which values should be filtered out
-/// \param[out] out resulting datum
+/// \param[in] options configures null_selection_behavior
+/// \param[out] out resulting array
 ARROW_EXPORT
-Status Filter(FunctionContext* ctx, const Datum& values, const Datum& filter, Datum* out);
+Status Filter(FunctionContext* ctx, const Datum& values, const Datum& filter,
+              FilterOptions options, Datum* out);
 
 /// \brief BinaryKernel implementing Filter operation
 class ARROW_EXPORT FilterKernel : public BinaryKernel {
  public:
-  explicit FilterKernel(const std::shared_ptr<DataType>& type) : type_(type) {}
+  explicit FilterKernel(std::shared_ptr<DataType> type, FilterOptions options)
```

Comment:
Should the constructor be public if the user is supposed to instantiate a subclass using the factory function?

Reply:
That's a good point. I'll move it into `protected:`